src/OrleansSQLUtils/CreateOrleansTables_MySql.sql

```diff
@@ -248,14 +328,156 @@ BEGIN
 	SET _ROWCOUNT = ROW_COUNT();
 
 	IF _ROWCOUNT = 0
-	THEN 
+	THEN
 		ROLLBACK;
 	ELSE
 		COMMIT;
 	END IF;
 	SELECT _ROWCOUNT;
 END$$
 
+CREATE PROCEDURE ClearStorage
```

Comment:
There is duplication and this is a fair point. I examine this part futher. though it looks like I can't use the same constructs directly.

This is not evident on the surface (which means I should spell it out in more explicit terms) is that the key isn't these two hashed integers alone. The actual keys are `GrainIdString` and `GrainTypeString`. Notes:

* There can be hash collisions, so these integers aren't unique, so it is not possible to enforce `PRIMARY KEY` or `UNIQUE` constraint on them.
* The constraint would need to be enforced on `GrainIdString` and `GrainTypeString`, but alas, they can become long and it looks like for generic grains with generic states the limit of `150` isn't enough. So in the general case these can be too thick to be used as keys and even if artificially constrained, the index would be thick (slow and resource intensive, impractical when there's a lot of data, require more maintenance).
* Enforcing `UNIQUE` or `PRIMARY` key likely means a sort order needs to be maintained, which increases resource contention and it wouldn't be possible on wide enough character columns.

Apart from artificially limiting the names and types of grains, the other concerns are that of performance. They don't matter much in small scale, but I think we should support state rows in the range of millions or even billions. Other thing is that the storage shouldn't choke on reads and updates and should handle surges of inserts too.

I don't intend to overwhelm with reference to authority, but allow me to produce some resources that describe this problem in more technical detail:
http://kejser.org/clustered-indexes-vs-heaps/
http://www.sqlpassion.at/archive/2014/04/15/an-ever-increasing-clustered-key-value-doesnt-scale/
http://kejser.org/boosting-insert-speed-by-generating-scalable-keys/
http://kejser.org/exploring-hash-functions-in-sql-server/
(for those out there with Oracle knowledge, "reverse index" might make sense)
[MySQL](https://dev.mysql.com/doc/refman/5.7/en/innodb-index-types.html) behaves slightly differently and depends on the engine used, but for what I know, is sufficiently similar. Both that and SQL Server maintain a special *row ID* that points to the actual record in this design (a pointer to data in code terms, I suppose). One difference is that while SQL Server looks like having improved performance, [MySQL with InnoDB](https://www.percona.com/blog/2013/10/18/innodb-scalability-issues-tables-without-primary-keys/) did suffer at least in some previous versions. Though there is still the problem on what should the primary key be enforced and wide keys is worse, it looks to me.

The take-away and noting the content of those resources is that there won't be range queries to this table, it is assumed `UPDATE` dominates and then `INSERT` should scale even on larger Orleans deployments and provide a fairly constant and stable operations and resource consumption on the storage side.